Add tags to samples. Closes #3640

martinlingstuyl opened the pull request:
Closes #3640

This PR adds tags to samples, creates a tags index page, and a page containing an list of available tags.

This really is a first draft. @pnp/cli-for-microsoft-365-maintainers, for a critical review.

Adam-it commented:
nice change 🤩. Thanks for opening this PR. We'll review it shortly 😉

Jwaegebaert commented:
Really love the implementation of the tags. I have nothing to add, the PR looks great from my perspective. Good job @martinlingstuyl 🚀

martinlingstuyl commented:
What about the chosen list of tags and the application of tags on the samples? Does that make sense?

garrytrinder commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,24 @@
+- [Guests](./tags#Guests)
+- [Lists](./tags#Lists)
+- [Migration](./tags#Migration)
+- [Miscelaneous](./tags#Miscelaneous)
```
Looks like a typo here

```suggestion
- [Miscellaneous](./tags#Miscellaneous)
```

waldekmastykarz replied:
This also needs to be updated in all samples

martinlingstuyl replied:
Yes,  but I'll remove the miscellaneous tag anyway

garrytrinder commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,24 @@
+- [Migration](./tags#Migration)
+- [Miscelaneous](./tags#Miscelaneous)
+- [Pages](./tags#Pages)
+- [Permission management](./tags#Permission-management)
```
How does this differ from `Security`?

garrytrinder reviewed (commented):
Great work @martinlingstuyl 👏

I'm wondering whether each tag needs a one-line description as to the purpose of the tag, this will make it easier for the contributor to understand which tags are relevant to their sample. It maybe be obvious with some tags, but not with others, so adding a little bit more clarity would go along way to helping folks make the right decision first time.

I'm on the fence with the `Miscellaneous` tag, I personally would not have one, they just become a catch-all tag.

martinlingstuyl commented:
> they just become a catch-all tag.

I expected this to come up 😂 I'll remove it.


I'll also add the one liners and merge security with permission management 👍

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,24 @@
+- [Guests](./tags#Guests)
+- [Lists](./tags#Lists)
+- [Migration](./tags#Migration)
+- [Miscelaneous](./tags#Miscelaneous)
```
What's the added value of such catch-all tag?

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,24 @@
+- [Sites](./tags#Sites)
+- [Social](./tags#Social)
+- [Tasks](./tags#Tasks)
+- [Usage](./tags#Usage)
```
Is this usage reporting?

martinlingstuyl replied:
I changed it to Reports...

martinlingstuyl commented:
Ok guys, I updated some things. Additional feedback very welcome!

Jwaegebaert commented:
Hey @martinlingstuyl, before we continue with the review. Could you resolve the conflicts this PR currently has. 😄

martinlingstuyl commented:
Will do!

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,21 @@
+
+Tag | Description
+-- | --
+[AI](./tags#AI) | Content dealing with AI related subjects. 
```
Let's simplify the descriptions and remove the _Content about/dealing/involving_. They're unnecessary and make it harder to read the descriptions.

waldekmastykarz replied:
Let's make all tags lowercase

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,21 @@
+
+Tag | Description
+-- | --
+[AI](./tags#AI) | Content dealing with AI related subjects. 
```
Not all descriptions are equally valuable. Eg. _AI - Content dealing with AI-related subjects_ is pretty self-explanatory, and doesn't add any value. On the other hand _Migration - Content about exporting, importing, syncing or moving data_ is better because it explains what migration is. Given that for all tags we're basically repeating the tag, we should revisit the stance of adding descriptions and if there's nothing to clarify, we should consider removing them altogether.

martinlingstuyl replied:
Ok, I agree. a lot of it is self explanatory, so I removed them. @garrytrinder

waldekmastykarz commented on docs/docs/sample-scripts/aad/analyze-users-haveibeenpwnd.md:
```diff
@@ -1,3 +1,9 @@
+---
+tags:  
+  - Privacy
```
Security would be more applicable here

waldekmastykarz commented on docs/docs/sample-scripts/aad/replace-membership-of-selected-groups.md:
```diff
@@ -1,3 +1,9 @@
+---
+tags:  
+  - Cleanup
```
Since we tagged `Replace an owner in a Microsoft 365 Group or Microsoft Team` with `security,users`, we should use the same tags here as well

waldekmastykarz commented on docs/docs/sample-scripts/flow/cancel-all-running-flow-runs.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:  
+  - Cleanup
```
This doesn't seem like a relevant tag to this script. To me, cleanup is more about removing resources. Cancelling running flows is more like management at scale/automation, which could be said about all scripts.

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,21 @@
+[Cleanup](./tags#Cleanup) | Content about removing things and cleaning up.
+[Files](./tags#Files) | Content about dealing with files.
+[Guests](./tags#Guests) | Content about working with external users.
+[Lists](./tags#Lists) | Content involving lists or libraries.
```
If we have tags for files, lists, sites, tasks, and users, shouldn't we have them for other entities as well, like flows, Teams teams, etc?

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,21 @@
+[Guests](./tags#Guests) | Content about working with external users.
+[Lists](./tags#Lists) | Content involving lists or libraries.
+[Migration](./tags#Migration) | Content about exporting, importing, syncing or moving data.
+[Failures](./tags#Failures) | Content dealing with problems and failures. 
```
This sounds like an odd tag. I can't imagine when I'd use it

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,21 @@
+[Privacy](./tags#Privacy) | Content about privacy related subjects.
+[Provisioning](./tags#Provisioning) | Content about creating structures or content programmatically.
+[Reports](./tags#Reports) | Content about getting lists of information for reporting purposes.
+[Security](./tags#Security) | Content about security and permissions.
```
Permissions are a part of security, so we should remove `security` from the description, because it's repeating the tag, without explaining what we actually mean.

waldekmastykarz commented on docs/docs/sample-scripts/spo/copy-listitems-sharepointlist.md:
```diff
@@ -1,3 +1,9 @@
+---
+tags:
+  - Lists
```
Shouldn't this be `list items`?

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,21 @@
+[Lists](./tags#Lists) | Content involving lists or libraries.
+[Migration](./tags#Migration) | Content about exporting, importing, syncing or moving data.
+[Failures](./tags#Failures) | Content dealing with problems and failures. 
+[Privacy](./tags#Privacy) | Content about privacy related subjects.
```
I'd say that `privacy` is pretty specific, and suggest we use `compliance` instead which is broader and allows us to reuse the tag across more samples.

waldekmastykarz commented on docs/docs/sample-scripts/spo/disable-tenant-wide-extension.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Provisioning
```
Shouldn't this be `cleanup`?

waldekmastykarz commented on docs/docs/sample-scripts/spo/insert-sp-library-pictures-into-word.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Lists
```
`files` would be more applicable here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-all-checked-out-files.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Files
```
We should add `reporting` here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-all-files-specific-name.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Files
```
We should add `reporting` here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-all-items-with-unique-permissions.md:
```diff
@@ -1,3 +1,9 @@
+tags:
+  - Security
+  - Lists
+---
```
We should add `reporting` here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-externalusers-in-sitegroups.md:
```diff
@@ -1,3 +1,9 @@
+tags:
+  - Guests  
+  - Users
+---
```
We should add `reporting` here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-site-app-catalogs.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Reports
+---
```
We should add `lists` here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-site-collection-lists.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Reports
+---
```
We should add `sites` and `lists` here

waldekmastykarz commented on docs/docs/sample-scripts/spo/list-site-externalusers.md:
```diff
@@ -1,3 +1,10 @@
+  - Security
+  - Users
+  - Guests
+---
```
We should add `reports`

waldekmastykarz commented on docs/docs/sample-scripts/spo/monitor-site-collection-storage-usage.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Reports
+---
```
We should add `sites`

waldekmastykarz commented on docs/docs/sample-scripts/spo/remove-api-permissions.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Security
+---
```
We should add `cleanup`

waldekmastykarz commented on docs/docs/sample-scripts/spo/remove-custom-themes.md:
```diff
@@ -1,3 +1,9 @@
+---
+tags:
+  - Provisioning
```
I don't think `provisioning` applies here as this script is more about cleaning up unneeded resources

waldekmastykarz commented on docs/docs/sample-scripts/spo/remove-orphaned-redirect-sites.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Cleanup
+---
```
We should add `sites`

waldekmastykarz commented on docs/docs/sample-scripts/spo/remove-pending-api-permission-requests.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Cleanup
+---
```
We should add `security`

waldekmastykarz commented on docs/docs/sample-scripts/teams/create-team-and-add-members-and-owners.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Users
+---
```
We should add `provisioning`

waldekmastykarz commented on docs/docs/sample-scripts/teams/get-all-teams-specific-user-is-part-of.md:
```diff
@@ -1,3 +1,8 @@
+---
+tags:
+  - Users
+---
```
We should add `reporting`

waldekmastykarz commented on docs/docs/sample-scripts/teams/govern-orphan-teams.md:
```diff
@@ -1,3 +1,9 @@
+tags:
+  - Users
+  - Security
+---
```
We should add `reporting`

waldekmastykarz commented:
I also noticed that many if not all samples have keywords at the end. Have we considered using them as tags as well?

martinlingstuyl commented:
Thanks for the valuable feedback @waldekmastykarz. 
I must say I largely ignored the keywords, as they often included for example the platform (SharePoint, Office 365) or a container term like Governance that I found questionable in its application.

I could check what keywords can be merged with tags, and delete all the keywords sections. That a good idea?

waldekmastykarz commented:
> I could check what keywords can be merged with tags, and delete all the keywords sections. That a good idea?

Good point! Let's do that 👍

martinlingstuyl commented:
Hi @waldekmastykarz, 
I revisited the list and all samples.

- I added some more tags and removed some others.
- I checked the keywords, merged them with tags if necessary, and removed them.
- I removed the descriptions.

Let's see what you think of it. 🤙

waldekmastykarz commented on docs/docs/tags.md:
```diff
@@ -0,0 +1,3 @@
+# Tags index
```
What do you think about merging this page with docs/docs/sample-scripts/index.md? That way, we don't introduce a new file that we don't show anywhere else, and when people come to the script samples section, they get an overview of the available scripts on a single page. This will also help us preserve the navigation on the left-hand side when browsing through the list of samples.

waldekmastykarz replied:
We've got a _Migration_ tag, that's uppercase _M_ and inconsistent with other tags, let's align it:

<img width="356" alt="image" src="https://user-images.githubusercontent.com/11164679/194704196-a33bed58-3416-4df3-85f6-1a0a658e09d8.png">

waldekmastykarz commented on docs/docs/tags.md:
```diff
@@ -0,0 +1,3 @@
+# Tags index
```
I noticed that titles of samples in the tags index, don't align with the actual titles:

<img width="275" alt="image" src="https://user-images.githubusercontent.com/11164679/194704220-b0bfe0de-6fac-4ed7-a216-828c2e2a85e7.png">

<img width="492" alt="image" src="https://user-images.githubusercontent.com/11164679/194704228-5cd68a64-b782-4e19-8855-d991d5ebe1dc.png">

This is confusing. Let's see why it's happening and if there's anything that we can do about it.

martinlingstuyl replied:
This is because of titles in the `nav` element in the `mkdocs.yml` file. 

They've been set up as: `- 'sample title': 'sample-scripts/spo/sample.md'`
This leads to differences with the page header in some cases. Apparently, the tags list also uses the `nav`-title.

I suggest we move to sample-scripts being defined as `- sample-scripts/spo/sample.md`. This will force mkdocs to use the first page header as a navigation title. Which in most cases translates exactly to what it is right now.

We might even do this for the rest of the navigation, but I'll leave that for now.

waldekmastykarz replied:
> I suggest we move to sample-scripts being defined as - sample-scripts/spo/sample.md. This will force mkdocs to use the first page header as a navigation title. Which in most cases translates exactly to what it is right now.

Great idea and a nice way to increase consistency. Let's create a new issue for it so that we don't lose track of it. We can leave this as-is and pick it up in the separate issue.

martinlingstuyl replied:
I already fixed it in this PR 🫢 maybe a bit too eager. Shall I withdraw those changes?

martinlingstuyl replied:
I've created an issue here:

https://github.com/pnp/cli-microsoft365/issues/3859

Would you think it better to move the changes I already made to sample-scripts items?

waldekmastykarz commented on docs/docs/_available-tags.md:
```diff
@@ -0,0 +1,29 @@
+# Available tags
```
Are we using this file anywhere? I can't seem to find any reference to it.

martinlingstuyl replied:
No that's correct. I'd thought it best to reference it in a wiki page where we describe how to contribute. We could also add a reference from the central tags md though. (The one I'll merge with index.md)

waldekmastykarz reviewed (changes requested):
Looks very nice! Let's do a few small adjustments before we merge it.

martinlingstuyl commented:
ok, I updated the code based on your comments @waldekmastykarz, don't forget to read my reply on the comment about mkdocs nav titles.

waldekmastykarz commented:
Hey @martinlingstuyl, looks great! I'm a bit on the fence when it comes to us maintaining a list of tags. Seems like our original idea was to have a list of tags from which contributors of new samples can choose when submitting samples. As such, it would make more sense for it to be in our contribution guide rather than docs, because that's where contributors will be looking at. We could also consider suggesting that contributors of new samples try to use existing tags as much as possible and introduce new tags only when we don't have anything else matching. That way we only have the automatically generated list from tags used in samples and don't have to keep another list up to date manually, which will help us save work and increase consistency.

waldekmastykarz reviewed (approved):
Works like a charm! 👏

waldekmastykarz commented:
Merged manually. Thank you! 👏

martinlingstuyl commented:
> Hey @martinlingstuyl, looks great! I'm a bit on the fence when it comes to us maintaining a list of tags. Seems like our original idea was to have a list of tags from which contributors of new samples can choose when submitting samples. As such, it would make more sense for it to be in our contribution guide rather than docs, because that's where contributors will be looking at. We could also consider suggesting that contributors of new samples try to use existing tags as much as possible and introduce new tags only when we don't have anything else matching. That way we only have the automatically generated list from tags used in samples and don't have to keep another list up to date manually, which will help us save work and increase consistency.

That's what we agreed on indeed. And you're right that they'd fit better in the contribution docs. I'll check it out and move the list of available tags there. 🤙
